synapse/types.py

```diff
@@ -400,6 +400,37 @@ def f2(m: Match[bytes]) -> bytes:
     return username.decode("ascii")
 
 
+MU = TypeVar("MU", bound="MXCUri")
+
+
+@attr.s(frozen=True, slots=True, auto_attribs=True, repr=False)
+class MXCUri:
+    """Represents a URI that points to a media resource in matrix.
+
+    MXC's take the form 'mxc://server_name/media_id'.
+    """
+
+    server_name: str
+    media_id: str
+
+    @classmethod
+    def from_str(cls: Type[MU], mxc_uri_str: str) -> MU:
+        try:
+            server_name, media_id = mxc_uri_str.split("/")[-2:]
+        except IndexError:
+            raise SynapseError(
+                400, f"Error processing malformed MXC uri: {mxc_uri_str}"
+            )
```

Comment:
This is probably fine, but raising SynapseError from the depths of the application has bitten us before and is not a pattern that I think we want to propagate.. Could you be persuaded to have this raise a ValueError instead?

Alternatively, we could have a constructor that returns Optional[MXCUri], with the expectation that users should check to see if the result is None to see if the validation was successful.